Fix 47 pleafletfinder

iparask opened the pull request:
Fixes #47 

Changes made in this Pull Request:
 - Adding new LeafletFInder class which analyzes in parallel a frame and reports the two leaflets.
 - Added a test file.
 - Add Timings functionality
 - Traverse over multiple frames.


This PR can be used as a place holder. I would like to initiate a discussion as to as the written code is based on the requirements. I would like to report two major differences from other cases. This code uses `Dask Bags` to execute instead of `delayed`. In addition, parallelization is done on a frame basis. This was the most efficient implementation we had in the paper (code linked in the respective issue).

Please review/ comment as necessary and code can grow.

Giannis

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+# Released under the GNU Public Licence, v2 or any higher version
+"""
+LeafletFInder Analysis tool --- :mod:`pmda.leaflet`
+=======================================
```
needs longer underscores or sphinx will barff

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+
+import numpy as np
+import dask.bag as db
+import networkx as nx
```
add networkx to the requirements in `setup.py` and in the travis setup – it gets installed with MDAnalysis but just be explicit

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+import networkx as nx
+
+import MDAnalysis as mda
+from sklearn.neighbors import BallTree
```
Is this required or can it be made optional or replaced by something from scipy?

At the moment the tests fail because sklearn is not installed.

We can decide to install sklearn if needed.

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+    Note
+    ----
+    At the moment, this class has far fewer features than the serial
+    version :class:`MDAnalysis.analysis.leaflet.LeafletFinder`.
```
Don't forget to add your paper.

iparask replied:
I did! Thank you for reminding me! I would have forgotten. :)

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+
+        Attributes
+        ----------
+        _results : list
```
If users are not supposed to use it (underscore prefixed) then don't document.

If users should be able to work with it, don't use underscore.

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,266 @@
+        else:
+            train = np.vstack([window[0],window[1]])
+            test  = np.vstack([window[0],window[1]])
+        tree = BallTree(train, leaf_size=40)
```
Can this be replaced with something from scipy and we use sklearn if available?

iparask commented:
The most efficient one uses BallTrees. It also reduces the memory footprint a lot. Let me include KDTrees instead, which are part of Scipy. Otherwise we will have to go with cdist or something similar.
@orbeckst, I missed the travis output.

orbeckst commented on pmda/test/test_leaflet.py:
```diff
@@ -0,0 +1,40 @@
+                                   assert_array_almost_equal)
+
+import MDAnalysis
+from MDAnalysisTests.datafiles import (PSF, DCD)
```
These datafiles do not include a membrane. 

Use the MDAnalysis tests for a single frame https://github.com/MDAnalysis/mdanalysis/blob/edb1643c0a78f7e5447aa3ccc02712e4205ff073/testsuite/MDAnalysisTests/analysis/test_leaflet.py#L30 to compare results and the files https://github.com/MDAnalysis/mdanalysis/blob/edb1643c0a78f7e5447aa3ccc02712e4205ff073/testsuite/MDAnalysisTests/analysis/test_rdf_s.py#L31 contain a trajectory with a membrane protein in a membrane.

iparask replied:
May I ask which atoms to select from the large trajectory? Thank you!

orbeckst replied:
Most likely "name P*" – sorry, can't check right now.

orbeckst replied:
@VOD555 what kind of lipids are in these trajectories and how do you select the headgroup phosphorous?

VOD555 replied:
There are POPG and POPE lipids in these trajectories, and you can use 
`u.select_atoms('name P and resname POPG')`

orbeckst commented:
Is BallTree a lot better? 

MDAnalysis also recently got [capped_distance](https://www.mdanalysis.org/mdanalysis/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.capped_distance) which could perhaps be used instead. We now use this for neighborsearching. Importantly, it takes periodicity into account.

@richardjgowers can you comment on `capped_distance()` vs `BallTree()` – did this come up during @ayushsuhane 's benchmarking during GSOC?

iparask commented:
@orbeckst @richardjgowers I have a [wiki}(https://github.com/radical-experiments/midas_exps/wiki/giannis_2_28_2017) page at least with some information for KD-Trees and BallTrees. Also, I remember @drelu had created a plot comparing different trees, I will put it here as soon as I find it.

iparask commented:
Here is Andre's experiment [link](https://github.com/radical-papers/midas-paper/blob/master/2017-TaskParallel-icpp/experiments/leaflet/distance/tree/cdist_tree.pdf)

I also found this benchmarking [blog](https://jakevdp.github.io/blog/2013/04/29/benchmarking-nearest-neighbor-searches-in-python/)

orbeckst commented:
Currently only as simple PEP8 check fails. 

You can run the PEP8 checks locally, too (together with coverage):
```
pytest  --pep8 -v --cov pmda  pmda
```
(needs pytest-pep8 and pytest-cov).

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        # Partition the data based on a 2-dimensional partitioning
+        for i in range(1, matrix_size + 1, part_size):
+            for j in range(1, matrix_size + 1, part_size):
+                arraged_coord.append(([atoms[i - 1:i - 1 + part_size],
```
If I understand this correctly, you are partitioning a (M, M) matrix A into sub-blocks. A_{ij} will represent a contact.

Because we evaluate the atomgroup with itself, this is symmetric, A_ij = A_ji. Couldn't you save on the order of M^2/2 evaluations by only taking the diagonal and lower triangle blocks?

orbeckst replied:
Did you have a look at optimizing this?

If it should be optimized, add it to the #72 issue.

For this PR we can put the code as it was for the paper.

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        comp = [g for g in subgraphs]
+        return comp
+
+    def _single_frame(self, scheduler_kwargs, n_blocks, cutoff=15.0):
```
Why `n_blocks` – it's not used at the moment. Omit?

(Unless you immediately want to parallelize over blocks, too)

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+            number of tasks to start, if `-1` use number of logical cpu cores.
+            This argument will be ignored when the distributed scheduler is
+            used
+        n_blocks : int, optional
```
Comment that this is currently ignore: no parallelization over trajectory blocks. Or are you already working on adding the parallelization over frames, too?

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        if n_jobs == -1:
+            n_jobs = cpu_count()
+
+        if n_blocks is None:
```
COuld be removed if `n_blocks` not used.

iparask replied:
It is being used to produce the correct number of tasks. Because it is optional, if it is not set we set it manually based on the number of jobs. Now that I think about it `n_jobs` is not really needed since I parallelize based on the number of blocks.

orbeckst commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        if scheduler == multiprocessing:
+            scheduler_kwargs['num_workers'] = n_jobs
+        
+        start, stop, step = self._trajectory.check_slice_indices(
```
none of the start/stop/step things are needed (unless you want to parallelize over frames)

If not needed, please remove to declutter.

iparask replied:
I do not parallelize over frames, but there is a `for loop` that goes over the frames

orbeckst replied:
ok, I missed that

iparask commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+    [Michaud-Agrawal2011]_.
+    The parallelization is done based on:
+
+    "Task-parallel Analysis of Molecular Dynamics Trajectories",
```
Okay!

iparask commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        else:
+            train = np.vstack([window[0], window[1]])
+            test = np.vstack([window[0], window[1]])
+        tree = cKDTree(train, leafsize=40)
```
What do you mean by PBC?

richardjgowers replied:
https://en.wikipedia.org/wiki/Periodic_boundary_conditions

like Pacman geometry.  Particles are simulated within a finite box of size [Lx, Ly, Lz], Particles near x=0 and x=Lx are actually very close to each other.

So the leaflet is actually an infinite sheet and has no "edge".  Without considering periodic boundaries you'll identify the leaflets, but you'll miss some of the edges in the graph representation, which might be important for some analysis

iparask replied:
Okay, I see. May I suggest the following. Have a working implementation without taking into account PBCs and without breaking the trajectory into multiple blocks. As soon as that is done and this PR can be accepted, merge it so that we do not miss the release this is intended for. 

Next version of the implementation will take into account PBCs and also try to see how to parallelize over trajectory blocks.

@orbeckst , @richardjgowers  do you agree with that?

orbeckst replied:
I am ok with an initial version that does not do PBC if
- the docs mention this
- we have an issue open for it

Breaking trajectory into blocks can also wait in my opinion.

iparask commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+        self._atomgroup = atomgroup
+        self._results = list()
+
+    def _find_parcc(self, data, cutoff=15.0):
```
Done.

iparask commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+
+        # partial connected components
+
+        subgraphs = nx.connected_components(graph)
```
If I understand your comment correctly, you are suggesting to change from netwrokx to the function MDAnalysis implements, right?

When I started the implementation of the Leaflet Finder for my paper I used the class that exists in MDAnalysis ([link](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/leaflet.py)). There networkx is being used. In addition, I was very cautious when I built my version to use highly used python packages for things I did not want to reimplement.

Did you reimplement a Connected component calculator or did that pre-exist Leaflet Finder?

Also, I am extremely bad with abbreviations and I have to google them all the time. Can you expand iirc?

orbeckst replied:
@richardjgowers did you benchmark `find_fragments()` against networkx?

orbeckst replied:
(Also, we should update the serial LeafletFinder with capped_distances and find_fragments...)

orbeckst replied:
I suggest we use networkx for the initial version because this is what was described in the paper. 

We can then open an new issue to optimize parallel leafletfinder and add PBC.

orbeckst commented on setup.py:
```diff
@@ -51,7 +51,9 @@
-        'joblib',  # cpu_count func currently
+        'joblib',
+        'networkx',
+        'scipy',  # cpu_count func currently
```
this comment belongs to joblib

iparask replied:
Done

iparask commented on pmda/leaflet.py:
```diff
@@ -0,0 +1,281 @@
+    ----
+    At the moment, this class has far fewer features than the serial
+    version :class:`MDAnalysis.analysis.leaflet.LeafletFinder`.
+
```
Done

orbeckst commented:
@iparask there's still an import of BallTree somewhere:
```
==================================== ERRORS ====================================
__________________ ERROR collecting pmda/test/test_leaflet.py __________________
ImportError while importing test module '/home/travis/build/MDAnalysis/pmda/pmda/test/test_leaflet.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
pmda/test/test_leaflet.py:10: in <module>
    from pmda import leaflet
pmda/leaflet.py:28: in <module>
    from sklearn.neighbors import BallTree
E   ModuleNotFoundError: No module named 'sklearn'
```

Once the tests pass and you need a final review, please ping me. Thanks!

iparask commented:
@orbeckst, this is a request for a final review. I have added all parts as asked. I believe it is as close as I can see it getting. Travis fails with warnings mainly because I have overridden and changed the methods of the base class in terms of input parameters.

orbeckst commented:
We'll have to fix the pylint warnings or exclude them; if we merge it like this, all future CI will fail.

Run
```
pylint pmda
```
locally while testing. I get:

```(pmda) oliver$ pylint pmda
Using config file /Volumes/Data/oliver/Biop/Projects/Methods/MDAnalysis/pmda/.pylintrc
************* Module pmda.leaflet
pmda/leaflet.py:73: [W0231(super-init-not-called), LeafletFinder.__init__] __init__ method from base class 'ParallelAnalysisBase' is not called
pmda/leaflet.py:115: [E1102(not-callable), LeafletFinder._find_connected_components] cKDTree is not callable
pmda/leaflet.py:159: [W0221(arguments-differ), LeafletFinder._single_frame] Parameters differ from overridden '_single_frame' method
pmda/leaflet.py:192: [W1619(old-division), LeafletFinder._single_frame] division w/o __future__ statement
pmda/leaflet.py:229: [W0221(arguments-differ), LeafletFinder.run] Parameters differ from overridden 'run' method

-----------------------------------
Your code has been rated at 9.77/10
```

Most of these warnings look fixable to me
- *pmda/leaflet.py:115: [E1102(not-callable), LeafletFinder._find_connected_components] cKDTree is not callable* — ???
- *pmda/leaflet.py:192: [W1619(old-division), LeafletFinder._single_frame] division w/o __future__ statement* — add the future import or replace with `//`
- *pmda/leaflet.py:159: [W0221(arguments-differ), LeafletFinder._single_frame] Parameters differ from overridden '_single_frame' method* — I think you still need to be able to accept all arguments that the standard `_single_frame()` could and just ignore them.
- *pmda/leaflet.py:229: [W0221(arguments-differ), LeafletFinder.run] Parameters differ from overridden 'run' method* — We might have decided on the different call signature during the review process. However, if at all possible we should keep the call generic signature intact. 

Disable the other pylint warning in code by bracketing the code with `# pylint: disable=...`/`# pylint: enable=...` blocks; see for example https://github.com/MDAnalysis/pmda/blob/3d78fc7c90e19cdc72932412b05c4c5d2d567fe7/pmda/rdf.py#L90

orbeckst commented:
On a second look, the mismatching call signatures might be a bit more difficult.... start with disabling the warnings in the code. I'll then comment in a code review.

(One way or another, the linter has to pass.)

orbeckst commented on CHANGELOG:
```diff
@@ -21,6 +21,7 @@ Enhancements
   * add add timing for _conclude and _prepare (Issue #49)
   * add parallel particle-particle RDF calculation module pmda.rdf (Issue #41)
   * add readonly_attributes context manager to ParallelAnalysisBase
+  * add parallel implementation of Leaflet Finder (Issue #47)
```
add your GitHub name at the top of 0.2.0!

add yourself to AUTHORS and the`authors` line in doc/conf.py

iparask replied:
Okay I will!

iparask commented:
Can you help me read the output of the codecov bot you included? I think that if we can test using the distributed scheduler the coverage will increase.

orbeckst commented:
@ianmkenney in PR #66  we will automatically run all possible parallelizations so no need to do this manually here.

Just add yourself as an author and we're done here.

orbeckst commented:
Thank you @iparask , congratulations to your first contribution!
